# First data explorations & messy transformations

In [122]:
import matplotlib.pyplot as plt
import seaborn
seaborn.set()  # this makes my plots pretty
%matplotlib inline

import numpy as np
import pandas as pd

from sklearn.ensemble import RandomForestClassifier
from sklearn import cross_validation as cv

from skll import metrics
from sklearn.metrics import make_scorer

import data_utils

In [108]:
train = pd.read_csv('../../data/train.csv').drop(['Id'], axis=1)
test = pd.read_csv('../../data/test.csv')

In [109]:
train = pd.get_dummies(train, columns=['Product_Info_2'])
test = pd.get_dummies(test, columns=['Product_Info_2'])

In [ ]:
# fill NaN values for Familiy_Hist_2-5 with means
for i in range(2, 6):
    col = 'Family_Hist_%i' % i
    train[col] = train[col].fillna(train[col].mean())
    test[col] = test[col].fillna(test[col].mean())
# and everything else with 0
train = train.fillna(0)
test = test.fillna(0)

## Train a basic Random Forest

In [103]:
kappa_scorer = make_scorer(metrics.kappa, weights='quadratic')

In [155]:
X_train = train.drop(['Response'], axis=1).values
y_train = train.Response.values
rfr_params = {'n_estimators': 300, 'random_state': 42, 'n_jobs': -1}
model = RandomForestClassifier(**rfr_params)
folds = cv.KFold(n=len(y_train), n_folds=5, shuffle=True, random_state=42)
scores = cv.cross_val_score(model, X_train, y_train, cv=folds, scoring=kappa_scorer, n_jobs=-1)
print(scores)
print(scores.mean(), scores.std())

[ 0.49593096  0.50901067  0.5108705   0.49834324  0.49726198]
(0.50228346962076531, 0.0063259403571169865)


### Predict testdata

In [158]:
X_test = test.drop(['Id'], axis=1).values
rfr_params = {'n_estimators': 300, 'random_state': 42, 'n_jobs': -1}
model = RandomForestClassifier(**rfr_params)
model.fit(X_train, y_train)
y_pred = model.predict(X_test)

In [133]:
data_utils.write_submission(y_pred, test)

### Scores

**CV**
[ 0.49593096  0.50901067  0.5108705   0.49834324  0.49726198] (0.50228346962076531, 0.0063259403571169865)

**LB**
0.51081

### Feature importance (RandomForest)
List the 50 most important features (acording to RF)

In [160]:
imp = pd.Series(model.feature_importances_, name='Importance')
names = pd.Series(train.drop(['Response'], axis=1).columns, name='Feature')
pd.concat([names, imp], axis=1).sort_values('Importance', ascending=False)[:50]

Feature  Importance
9                    BMI    0.091645
8                     Wt    0.066192
6                Ins_Age    0.040889
2         Product_Info_4    0.039900
10     Employment_Info_1    0.036277
7                     Ht    0.031799
37     Medical_History_2    0.030827
34         Family_Hist_4    0.029889
36     Medical_History_1    0.029585
15     Employment_Info_6    0.029077
39     Medical_History_4    0.027248
33         Family_Hist_3    0.025160
32         Family_Hist_2    0.024280
27   Insurance_History_5    0.021863
18         InsuredInfo_3    0.021701
50    Medical_History_15    0.021146
35         Family_Hist_5    0.019225
91    Medical_Keyword_15    0.016791
11     Employment_Info_2    0.016558
58    Medical_History_23    0.015648
13     Employment_Info_4    0.009291
31         Family_Hist_1    0.009238
21         InsuredInfo_6    0.009192
79     Medical_Keyword_3    0.008759
29   Insurance_History_8    0.008676
76    Medical_History_41    0.008111
26   Insurance_History_4    0.007469
44     Medical_History_9    0.007333
16         InsuredInfo_1    0.007142
141    Product_Info_2_D3    0.007123
64    Medical_History_29    0.006908
4         Product_Info_6    0.006439
71    Medical_History_36    0.006397
74    Medical_History_39    0.006326
69    Medical_History_34    0.006199
28   Insurance_History_7    0.005895
60    Medical_History_25    0.005756
51    Medical_History_16    0.005720
59    Medical_History_24    0.005659
19         InsuredInfo_4    0.005654
1         Product_Info_3    0.005633
23   Insurance_History_1    0.005605
139    Product_Info_2_D1    0.005596
142    Product_Info_2_D4    0.005478
61    Medical_History_26    0.005405
48    Medical_History_13    0.005352
140    Product_Info_2_D2    0.004879
65    Medical_History_30    0.004814
30   Insurance_History_9    0.004757
43     Medical_History_8    0.004711

## histograms and scatterplots
(deactivated)